src/main/java/com/google/cloud/dataproc/templates/main/DataProcTemplate.java

```diff
@@ -15,86 +15,116 @@
  */
 package com.google.cloud.dataproc.templates.main;
 
+import com.google.cloud.dataproc.templates.BaseTemplate;
 import com.google.cloud.dataproc.templates.BaseTemplate.TemplateName;
+import com.google.cloud.dataproc.templates.GeneralTemplate;
 import com.google.cloud.dataproc.templates.databases.SpannerToGCS;
 import com.google.cloud.dataproc.templates.gcs.GCStoBigquery;
 import com.google.cloud.dataproc.templates.hive.HiveToBigQuery;
 import com.google.cloud.dataproc.templates.hive.HiveToGCS;
 import com.google.cloud.dataproc.templates.pubsub.PubSubToBQ;
 import com.google.cloud.dataproc.templates.s3.S3ToBigQuery;
+import com.google.cloud.dataproc.templates.util.PropertyUtil;
 import com.google.cloud.dataproc.templates.word.WordCount;
+import com.google.common.collect.ImmutableMap;
+import java.util.Map;
+import java.util.Properties;
+import java.util.function.Function;
+import org.apache.commons.cli.BasicParser;
+import org.apache.commons.cli.CommandLine;
+import org.apache.commons.cli.CommandLineParser;
+import org.apache.commons.cli.Option;
+import org.apache.commons.cli.OptionBuilder;
+import org.apache.commons.cli.Options;
+import org.apache.commons.cli.ParseException;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 public class DataProcTemplate {
 
   private static final Logger LOGGER = LoggerFactory.getLogger(DataProcTemplate.class);
+  private static final String TEMPLATE_NAME = "template";
 
-  public static void main(String[] args) {
-    if (args.length < 1) {
-      throw new IllegalArgumentException(
-          "Minimum 1 arguments is required: " + "<template-name> Ex wordcount");
+  /**
+   * Parse command line arguments
+   *
+   * @param args command line arguments
+   * @return parsed arguments
+   */
+  public static CommandLine parseArguments(String... args) {
```

Comment:
Template name and extra properties are parsed here, any other arguments are considered remainders and passed on to the template factory.